internal/kibana/integrations.go

```diff
@@ -22,24 +25,65 @@ type IntegrationPackage struct {
 
 // AddIntegrationToPolicy adds an integration to policy
 func (c *Client) AddIntegrationToPolicy(ctx context.Context, packageDS PackageDataStream) error {
-	span, _ := apm.StartSpanOptions(ctx, "Adding integration to policy", "fleet.package.add-to-policy", apm.SpanOptions{
-		Parent: apm.SpanFromContext(ctx).TraceContext(),
-	})
-	defer span.End()
+	maxTimeout := time.Duration(utils.TimeoutFactor) * time.Minute
+	retryCount := 1
 
-	reqBody, err := json.Marshal(packageDS)
-	if err != nil {
-		return errors.Wrap(err, "could not convert policy-package (request) to JSON")
+	exp := utils.GetExponentialBackOff(maxTimeout)
+
+	addIntegrationFn := func() error {
+		span, _ := apm.StartSpanOptions(ctx, "Adding integration to policy", "fleet.package.add-to-policy", apm.SpanOptions{
+			Parent: apm.SpanFromContext(ctx).TraceContext(),
+		})
+		defer span.End()
+
+		reqBody, err := json.Marshal(packageDS)
+		if err != nil {
+			log.WithFields(log.Fields{
+				"elapsedTime": exp.GetElapsedTime(),
+				"err":         err,
+				"package":     packageDS,
+				"retry":       retryCount,
+			}).Warn("Could not convert policy-package (request) to JSON. Retrying")
+
+			retryCount++
```

Comment:
This is probably just me, but I don't understand how the retry logic is working here. Is this variable staying in scope and somehow being consumed by `backoff.Retry`? In looking at the documentation, I can't see any mention of this behavior so I wonder if perhaps https://pkg.go.dev/github.com/cenkalti/backoff/v4#WithMaxRetries should be used instead?

(Haven't had much coffee yet so possible that I'm missing something obvious.)

Reply:
Mmm, this is interesting because I hadn't explored that part of the API yet... The retry variable here is more or less cosmetic, as it's used to provide feedback in logs about what's the current state of the retry function. Apart from that, we use an ExponentialBackoff, which is another type: https://pkg.go.dev/github.com/cenkalti/backoff/v4#ExponentialBackOff, adding a non-linear wait time between following calls.

Reply:
It's interesting to me that it works at all like that. I would have thought that the variable went out of scope after each invocation of the retry. There must be something fancy happening in the retry library to preserve it. Thanks for the explanation!

Reply:
Because it's declared outside of the retry function, its value is preserved as a "global-to-the-function"